File: ObjcExample/ObjcExample/Demo/豆瓣/GKDBViewController.m
//
//  GKDBViewController.m
//  GKPageScrollViewObjc
//
//  Created by gaokun on 2020/12/17.
//  Copyright © 2020 gaokun. All rights reserved.
//

#import "GKDBViewController.h"
#import "GKPageSmoothView.h"
#import "GKDBListView.h"

@interface GKDBViewController ()<GKPageSmoothViewDataSource, GKPageSmoothViewDelegate, JXCategoryViewDelegate>

@property (nonatomic, strong) UIView *titleView;

@property (nonatomic, strong) GKPageSmoothView *smoothView;

@property (nonatomic, strong) UIImageView *headerView;

@property (nonatomic, strong) UIView    *segmentedView;

@property (nonatomic, strong) JXCategorySubTitleView *categoryView;

@property (nonatomic, strong) JXCategoryIndicatorAlignmentLineView *lineView;

@property (nonatomic, assign) BOOL isTitleViewShow;
@property (nonatomic, assign) CGFloat originAlpha;

@property (nonatomic, assign) CGFloat lastRatio;

@end

@implementation GKDBViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.gk_statusBarStyle = UIStatusBarStyleLightContent;
    self.gk_navBarAlpha = 0;
    self.gk_navBackgroundColor = GKColorRGB(123, 106, 89);
    self.gk_navTitle = @"电影";
    self.gk_navTitleColor = UIColor.whiteColor;
    
    [self.view addSubview:self.smoothView];
    [self.smoothView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view);
    }];
    
    self.categoryView.contentScrollView = self.smoothView.listCollectionView;
    [self.smoothView reloadData];
}

#pragma mark - GKPageSmoothViewDataSource
- (UIView *)headerViewInSmoothView:(GKPageSmoothView *)smoothView {
    return self.headerView;
}

- (UIView *)segmentedViewInSmoothView:(GKPageSmoothView *)smoothView {
    return self.segmentedView;
}

- (NSInteger)numberOfListsInSmoothView:(GKPageSmoothView *)smoothView {
    return self.categoryView.titles.count;
}

- (id<GKPageSmoothListViewDelegate>)smoothView:(GKPageSmoothView *)smoothView initListAtIndex:(NSInteger)index {
    GKDBListView *listView = [GKDBListView new];
    return listView;
}

#pragma mark - GKPageSmoothViewDelegate
- (void)smoothView:(GKPageSmoothView *)smoothView listScrollViewDidScroll:(UIScrollView *)scrollView contentOffset:(CGPoint)contentOffset {
    if (smoothView.isOnTop) return;
    
    // 导航栏显隐
    CGFloat offsetY = contentOffset.y;
    CGFloat alpha = 0;
    if (offsetY <= 0) {
        alpha = 0;
    }else if (offsetY > 60) {
        alpha = 1;
        [self changeTitle:YES];
    }else {
        alpha = offsetY / 60;
        [self changeTitle:NO];
    }
    self.gk_navBarAlpha = alpha;
}

- (void)smoothViewDragBegan:(GKPageSmoothView *)smoothView {
    if (smoothView.isOnTop) return;
    
    self.isTitleViewShow = (self.gk_navTitleView != nil);
    self.originAlpha = self.gk_navBarAlpha;
}

- (void)smoothViewDragEnded:(GKPageSmoothView *)smoothView isOnTop:(BOOL)isOnTop {
    // titleView已经显示，不作处理
    if (self.isTitleViewShow) return;
    
    if (isOnTop) {
        self.gk_navBarAlpha = 1.0f;
        [self changeTitle:YES];
    }else {
        self.gk_navBarAlpha = self.originAlpha;
        [self changeTitle:NO];
    }
}

- (void)changeTitle:(BOOL)isShow {
    if (isShow) {
        if (self.gk_navTitle == nil) return;
        self.gk_navTitle = nil;
        self.gk_navTitleView = self.titleView;
    }else {
        if (self.gk_navTitleView == nil) return;
        self.gk_navTitle = @"电影";
        self.gk_navTitleView = nil;
    }
}

#pragma mark - JXCategoryViewDelegate
- (void)categoryView:(JXCategoryBaseView *)categoryView didClickSelectedItemAtIndex:(NSInteger)index {
    [self.smoothView showingOnTop];
}

#pragma mark - 懒加载
- (UIView *)titleView {
    if (!_titleView) {
        _titleView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width - 100, 44.0f)];
        
        UIImage *image = [UIImage imageNamed:@"db_title"];
        UIImageView *imgView = [[UIImageView alloc] initWithImage:image];
        imgView.frame = CGRectMake(0, 0, 44.0f * image.size.width / image.size.height, 44.0f);
        [_titleView addSubview:imgView];
    }
    return _titleView;
}

- (GKPageSmoothView *)smoothView {
    if (!_smoothView) {
        _smoothView = [[GKPageSmoothView alloc] initWithDataSource:self];
        _smoothView.delegate = self;
        _smoothView.ceilPointHeight = GK_STATUSBAR_NAVBAR_HEIGHT;
        _smoothView.bottomHover = YES;
        _smoothView.allowDragBottom = YES;
        _smoothView.allowDragScroll = YES;
        // 解决与返回手势滑动冲突
        _smoothView.listCollectionView.gk_openGestureHandle = YES;
    }
    return _smoothView;
}

- (UIImageView *)headerView {
    if (!_headerView) {
        UIImage *image = [UIImage imageNamed:@"douban"];
        _headerView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, image.size.width, image.size.height)];
        _headerView.image = image;
    }
    return _headerView;
}

- (UIView *)segmentedView {
    if (!_segmentedView) {
        _segmentedView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, 60)];
        _segmentedView.backgroundColor = [UIColor whiteColor];
        [_segmentedView addSubview:self.categoryView];
        
        UIView *topView = [UIView new];
        topView.backgroundColor = [UIColor lightGrayColor];
        topView.layer.cornerRadius = 3;
        topView.layer.masksToBounds = YES;
        [_segmentedView addSubview:topView];
        [topView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(self->_segmentedView).offset(5);
            make.centerX.equalTo(self->_segmentedView);
            make.width.mas_equalTo(60);
            make.height.mas_equalTo(6);
        }];
    }
    return _segmentedView;
}

- (JXCategorySubTitleView *)categoryView {
    if (!_categoryView) {
        _categoryView = [[JXCategorySubTitleView alloc] initWithFrame:CGRectMake(0, 10, self.view.frame.size.width, 40)];
        _categoryView.backgroundColor = UIColor.whiteColor;
        _categoryView.averageCellSpacingEnabled = NO;
        _categoryView.contentEdgeInsetLeft = 16;
        _categoryView.delegate = self;
        _categoryView.titles = @[@"影评", @"讨论"];
        _categoryView.titleFont = [UIFont systemFontOfSize:16];
        _categoryView.titleColor = UIColor.grayColor;
        _categoryView.titleSelectedColor = UIColor.blackColor;
        _categoryView.subTitles = @[@"342", @"2004"];
        _categoryView.subTitleFont = [UIFont systemFontOfSize:11];
        _categoryView.subTitleColor = UIColor.grayColor;
        _categoryView.subTitleSelectedColor = UIColor.grayColor;
        _categoryView.positionStyle = JXCategorySubTitlePositionStyle_Right;
        _categoryView.alignStyle = JXCategorySubTitleAlignStyle_Top;
        _categoryView.cellSpacing = 30;
        _categoryView.cellWidthIncrement = 0;
        _categoryView.ignoreSubTitleWidth = YES;
        
        JXCategoryIndicatorLineView *lineView = [JXCategoryIndicatorLineView new];
        lineView.indicatorColor = UIColor.blackColor;
        _categoryView.indicators = @[self.lineView];
        
        _categoryView.contentScrollView = self.smoothView.listCollectionView;
    }
    return _categoryView;
}

-  (JXCategoryIndicatorAlignmentLineView *)lineView {
    if (!_lineView) {
        _lineView = [JXCategoryIndicatorAlignmentLineView new];
        _lineView.indicatorColor = UIColor.blackColor;
    }
    return _lineView;
}

@end

File: SwiftExample/SwiftExample/Demo/GKPageSmoothView/豆瓣/GKDBViewController.swift
//
//  GKDBViewController.swift
//  GKPageScrollViewSwift
//
//  Created by gaokun on 2020/12/23.
//  Copyright © 2020 gaokun. All rights reserved.
//

import UIKit
import JXSegmentedView
import GKNavigationBarSwift
import GKPageSmoothView

class GKDBViewController: GKDemoBaseViewController {

    lazy var titleView: UIView = {
        let titleView = UIView(frame: CGRect(x: 0, y: 0, width: self.view.frame.size.width - 100, height: 44.0))
        
        let image = UIImage(named: "db_title")
        let imgView = UIImageView(image: image)
        imgView.frame = CGRect(x: 0, y: 0, width: 44.0 * image!.size.width / image!.size.height, height: 44.0)
        titleView.addSubview(imgView)
        return titleView
    }()
    
    lazy var smoothView: GKPageSmoothView = {
        let smoothView = GKPageSmoothView(dataSource: self)
        smoothView.delegate = self
        smoothView.ceilPointHeight = GK_STATUSBAR_NAVBAR_HEIGHT
        smoothView.isBottomHover = true
        smoothView.isAllowDragBottom = true
        smoothView.isAllowDragScroll = true
        smoothView.listCollectionView.gk_openGestureHandle = true
        return smoothView
    }()
    
    lazy var headerView: UIImageView = {
        let image = UIImage(named: "douban")
        let imgView = UIImageView(image: image)
        imgView.frame = CGRect(x: 0, y: 0, width: image!.size.width, height: image!.size.height)
        return imgView
    }()
    
    lazy var segmentedView: UIView = {
        let view = UIView(frame: CGRect(x: 0, y: 0, width: kScreenW, height: 60))
        view.backgroundColor = .white
        view.addSubview(self.titleSegmentedView)
        
        let topView = UIView()
        topView.backgroundColor = .lightGray
        topView.layer.cornerRadius = 3.0
        topView.layer.masksToBounds = true
        view.addSubview(topView)
        topView.snp.makeConstraints { (make) in
            make.top.equalTo(view).offset(5)
            make.centerX.equalTo(view)
            make.width.equalTo(60)
            make.height.equalTo(6)
        }
        return view
    }()
    
    var dataSource = JXSegmentedNumberDataSource()
    
    lazy var titleSegmentedView: JXSegmentedView = {
        dataSource.titles = ["影评", "讨论"]
        dataSource.numbers = [342, 200]
        dataSource.isItemSpacingAverageEnabled = false
        dataSource.numberBackgroundColor = UIColor.clear
        dataSource.numberTextColor = UIColor.gray
        dataSource.numberFont = UIFont.systemFont(ofSize: 11)
        dataSource.numberOffset = CGPoint(x: 10, y: 6)
        dataSource.titleNormalFont = UIFont.systemFont(ofSize: 16)
        dataSource.titleNormalColor = UIColor.gray
        dataSource.titleSelectedColor = UIColor.black
        dataSource.itemSpacing = 50
        
        var segmentedView = JXSegmentedView(frame: CGRect(x: 0, y: 10, width: kScreenW, height: 40))
        segmentedView.delegate = self
        segmentedView.dataSource = dataSource
        segmentedView.contentEdgeInsetLeft = 16
        segmentedView.backgroundColor = .white
        
        let lineView = JXSegmentedIndicatorLineView()
        lineView.lineStyle = .normal
        lineView.indicatorColor = .black
        segmentedView.indicators = [lineView]
        return segmentedView
    }()
    
    var isTitleViewShow = false
    var originAlpha: CGFloat = 0
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.gk_statusBarStyle = .lightContent
        self.gk_navBarAlpha = 0
        self.gk_navBackgroundColor = UIColor(red: 123.0 / 255.0, green: 106.0 / 255.0, blue: 89.0 / 255.0, alpha: 1.0)
        self.gk_navTitle = "电影"
        self.gk_navTitleColor = .white
        
        self.view.addSubview(self.smoothView)
        self.smoothView.snp.makeConstraints { (make) in
            make.edges.equalTo(self.view)
        }
        self.titleSegmentedView.contentScrollView = self.smoothView.listCollectionView
        self.smoothView.reloadData()
    }
    
    func changeTitle(isShow: Bool) {
        if isShow {
            if (self.gk_navTitle == nil) {return}
            self.gk_navTitle = nil
            self.gk_navTitleView = self.titleView
        }else {
            if (self.gk_navTitleView == nil) {return}
            self.gk_navTitle = "电影"
            self.gk_navTitleView = nil
        }
    }
    
    func attributedTextFrom(title: String, count: NSNumber, isSelected: Bool) -> NSAttributedString {
        let countString = String(format: "%@", count)
        let allString = String(format: "%@%@", title, countString)
        
        var tintColor: UIColor = .gray
        if isSelected {
            tintColor = .black
        }
        
        let attributedText = NSMutableAttributedString(string: allString, attributes: [NSAttributedString.Key.font: UIFont.systemFont(ofSize: 16), NSAttributedString.Key.foregroundColor: tintColor])
        let titleRange = (title as NSString).range(of: title)
        let countRange = (allString as NSString).range(of: countString)
        attributedText.addAttribute(.font, value: UIFont.systemFont(ofSize: 12), range: countRange)
        attributedText.addAttribute(.foregroundColor, value: UIColor.gray, range: countRange)
        
        let fontRatio: CGFloat = 0.96
        
        attributedText.addAttribute(.baselineOffset, value: 0, range: titleRange)
        attributedText.addAttribute(.baselineOffset, value: fontRatio * (16 - 12), range: countRange)
        return attributedText
    }
}

extension GKDBViewController: GKPageSmoothViewDataSource {
    func headerView(in smoothView: GKPageSmoothView) -> UIView {
        return self.headerView
    }
    
    func segmentedView(in smoothView: GKPageSmoothView) -> UIView {
        return self.segmentedView
    }
    
    func numberOfLists(in smoothView: GKPageSmoothView) -> Int {
        return dataSource.titles.count
    }
    
    func smoothView(_ smoothView: GKPageSmoothView, initListAtIndex index: Int) -> GKPageSmoothListViewDelegate {
        return GKDBListView()
    }
}

extension GKDBViewController: GKPageSmoothViewDelegate {
    func smoothViewListScrollViewDidScroll(_ smoothView: GKPageSmoothView, scrollView: UIScrollView, contentOffset: CGPoint) {
        if (smoothView.isOnTop) { return }
        // 导航栏显隐
        let offsetY = contentOffset.y
        var alpha: CGFloat = 0
        if offsetY <= 0 {
            alpha = 0
        }else if offsetY > 60 {
            alpha = 1
            changeTitle(isShow: true)
        }else {
            alpha = offsetY / 60.0
            changeTitle(isShow: false)
        }
        self.gk_navBarAlpha = alpha
    }
    
    func smoothViewDragBegan(_ smoothView: GKPageSmoothView) {
        if smoothView.isOnTop {return}
        self.isTitleViewShow = (self.gk_navTitleView != nil)
        self.originAlpha = self.gk_navBarAlpha
    }
    
    func smoothViewDragEnded(_ smoothView: GKPageSmoothView, isOnTop: Bool) {
        // titleView已经显示，不作处理
        if self.isTitleViewShow {return}
        if isOnTop {
            self.gk_navBarAlpha = 1.0
            changeTitle(isShow: true)
        }else {
            self.gk_navBarAlpha = 0.0
            changeTitle(isShow: false)
        }
    }
}

extension GKDBViewController: JXSegmentedViewDelegate {
    func segmentedView(_ segmentedView: JXSegmentedView, didClickSelectedItemAt index: Int) {
        self.smoothView.showingOnTop()
    }
}
